### PR TITLE
Delete various end symbols

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
For various headers we need to be able to refer to ends of various sections within the compiler. We were generating named symbols for these, but they can also be references as "beginning symbol + delta". The end symbol often confuses debuggers and disassemblers and instead of saying "this symbol is FooSection+0x123", it will say "BarSectionEnd+0x123" and that's a useless information.

Delete the named end symbols in favor of generating relocs with deltas.

Cc @dotnet/ilc-contrib

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReadyToRunHeaderNode.cs [hunk 1]
```diff
@@ -5,33 +5,31 @@
...
-            public readonly ISymbolNode EndSymbol;
         }
 
         private List<HeaderItem> _items = new List<HeaderItem>();
 
-        public void Add(ReadyToRunSectionType id, ObjectNode node, ISymbolNode startSymbol, ISymbolNode endSymbol = null)
+        public void Add(ReadyToRunSectionType id, ObjectNode node)
         {
-            _items.Add(new HeaderItem(id, node, startSymbol, endSymbol));
+            Debug.Assert(node is ISymbolDefinitionNode);
```
Comment (filipnavara) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReadyToRunHeaderNode.cs:31]: Out of curiosity, wouldn't it make sense to ensure this through generic constraint?

```csharp
public void Add<TObjectNode>(ReadyToRunSectionType id, TObjectNode node)
    where TObjectNode : ObjectNode, ISymbolDefinitionNode
```

**Review by filipnavara — COMMENTED** [review]
Generally looks okay to me.

Some formats, such as ELF, have the concept of symbols with sizes. It would be nice for someone to chime in whether that's of any use for this scenario or not. I think we emit the sizes for function symbols, and I am not sure whether it's allowed for other types of symbols or even beneficial in any way.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReadyToRunHeaderNode.cs (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReadyToRunHeaderNode.cs]: Yup, good point!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReadyToRunHeaderNode.cs [hunk 2]
```diff
@@ -117,6 +115,12 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
             return builder.ToObjectData();
         }
 
-        public override int ClassCode => -534800244;
+        protected internal override int Phase => (int)ObjectNodePhase.Late;
+        public override int ClassCode => 0x7db08464;
```
[the remark this answers is not in the document]
Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReadyToRunHeaderNode.cs:118]: It's fine to change this, it's used for ordering only.

**Comment by MichalStrehovsky** [comment]
> Some formats, such as ELF, have the concept of symbols with sizes.

Even if there's some special reloc for this on ELF, we'd need to emulate it elsewhere and emulating it is hard because we don't generally know the size (would need to call GetData to get the size and many things assume we're going to only ever call GetData twice - once with relocsOnly during dependency analysis, and second time in object emission). GetData can be quite expensive to retreive.

**Review by jkotas — APPROVED** [review]
LGTM!

**Comment by filipnavara** [comment]
> Even if there's some special reloc for this on ELF, we'd need to emulate it elsewhere...

I was thinking more in line that it may be useful to propagate the information to get a better debugging experience if the symbol size is used by the debugger...

**Comment by MichalStrehovsky** [comment]
> > Even if there's some special reloc for this on ELF, we'd need to emulate it elsewhere...
> 
> I was thinking more in line that it may be useful to propagate the information to get a better debugging experience if the symbol size is used by the debugger...

Possibly - but I'm not planning to experiment on whether it helps with anything because I don't know if there's any known issue with it right now.

**Pull request merged**